File: DslAdapter/src/main/java/com/angcyo/dsladapter/RDiffCallback.java
package com.angcyo.dsladapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;
import android.text.TextUtils;

import java.util.List;

public class RDiffCallback<T> extends DiffUtil.Callback {
    List<T> oldDatas;
    List<T> newDatas;
    RDiffCallback<T> mDiffCallback;

    public RDiffCallback() {
    }

    public RDiffCallback(List<T> oldDatas, List<T> newDatas, RDiffCallback<T> diffCallback) {
        this.oldDatas = oldDatas;
        this.newDatas = newDatas;
        mDiffCallback = diffCallback;
    }

    public static int getListSize(List list) {
        return list == null ? 0 : list.size();
    }

    @Override
    public int getOldListSize() {
        return getListSize(oldDatas);
    }

    @Override
    public int getNewListSize() {
        return getListSize(newDatas);
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        return super.getChangePayload(oldItemPosition, newItemPosition);
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return mDiffCallback.areItemsTheSame(oldDatas.get(oldItemPosition), newDatas.get(newItemPosition));
    }

    /**
     * 被DiffUtil调用，用来检查 两个item是否含有相同的数据
     * 这个方法仅仅在areItemsTheSame()返回true时，才调用。
     */
    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return mDiffCallback.areContentsTheSame(oldDatas.get(oldItemPosition), newDatas.get(newItemPosition));
    }

    /**
     * 重写此方法, 判断数据是否相等,
     * 如果item不相同, 会先调用 notifyItemRangeRemoved, 再调用 notifyItemRangeInserted
     */
    public boolean areItemsTheSame(@NonNull T oldData, @NonNull T newData) {
        Class<?> oldClass = oldData.getClass();
        Class<?> newClass = newData.getClass();
        if (oldClass.isAssignableFrom(newClass) || newClass.isAssignableFrom(oldClass)) {
            return true;
        }
        return TextUtils.equals(oldClass.getSimpleName(), newClass.getSimpleName());
    }

    /**
     * 重写此方法, 判断内容是否相等,
     * 如果内容不相等, 会调用notifyItemRangeChanged
     */
    public boolean areContentsTheSame(@NonNull T oldData, @NonNull T newData) {
        Class<?> oldClass = oldData.getClass();
        Class<?> newClass = newData.getClass();
        if (oldClass.isAssignableFrom(newClass) ||
                newClass.isAssignableFrom(oldClass) ||
                TextUtils.equals(oldClass.getSimpleName(), newClass.getSimpleName())) {
            return oldData.equals(newData);
        }
        return false;
    }
}

File: Adapter/src/main/java/com/angcyo/dsladapter/internal/RDiffCallback.kt
package com.angcyo.dsladapter.internal

import android.text.TextUtils
import androidx.recyclerview.widget.DiffUtil

open class RDiffCallback<T : Any>(
    val oldDataList: List<T>,
    val newDataList: List<T>,
    val itemDiffCallback: RItemDiffCallback<T>? = null
) : DiffUtil.Callback() {

    companion object {
        fun getListSize(list: List<*>?): Int {
            return list?.size ?: 0
        }
    }

    override fun getOldListSize(): Int {
        return getListSize(
            oldDataList
        )
    }

    override fun getNewListSize(): Int {
        return getListSize(
            newDataList
        )
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        return itemDiffCallback?.getChangePayload(
            oldDataList[oldItemPosition],
            newDataList[newItemPosition],
            oldItemPosition, newItemPosition
        )
    }

    override fun areItemsTheSame(
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean {
        return itemDiffCallback?.areItemsTheSame(
            oldDataList[oldItemPosition],
            newDataList[newItemPosition],
            oldItemPosition, newItemPosition
        ) ?: false
    }

    /**
     * 被DiffUtil调用，用来检查 两个item是否含有相同的数据
     * 这个方法仅仅在areItemsTheSame()返回true时，才调用。
     */
    override fun areContentsTheSame(
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean {
        return itemDiffCallback?.areContentsTheSame(
            oldDataList[oldItemPosition],
            newDataList[newItemPosition],
            oldItemPosition, newItemPosition
        ) ?: false
    }
}

interface RItemDiffCallback<T : Any> {
    fun getChangePayload(oldData: T, newData: T, oldItemPosition: Int, newItemPosition: Int): Any? {
        return null
    }

    /**
     * 重写此方法, 判断数据是否相等,
     * 如果item不相同, 会先调用 notifyItemRangeRemoved, 再调用 notifyItemRangeInserted
     */
    fun areItemsTheSame(
        oldData: T,
        newData: T,
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean {
        val oldClass: Class<*> = oldData.javaClass
        val newClass: Class<*> = newData.javaClass
        return TextUtils.equals(
            oldClass.simpleName,
            newClass.simpleName
        ) && oldItemPosition == newItemPosition
    }

    /**
     * 重写此方法, 判断内容是否相等,
     * 如果内容不相等, 会调用notifyItemRangeChanged
     */
    fun areContentsTheSame(
        oldData: T,
        newData: T,
        oldItemPosition: Int,
        newItemPosition: Int
    ): Boolean {
        val oldClass: Class<*> = oldData.javaClass
        val newClass: Class<*> = newData.javaClass
        return if (oldClass.isAssignableFrom(newClass) ||
            newClass.isAssignableFrom(oldClass) ||
            TextUtils.equals(oldClass.simpleName, newClass.simpleName)
        ) {
            oldData == newData
        } else false
    }
}